profiler/src/lib.rs

```diff
@@ -0,0 +1,67 @@
+//! A module for profiling canisters.
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+
+thread_local! {
+    static PROFILING: RefCell<BTreeMap<&'static str, u64>> = RefCell::new(BTreeMap::new());
+}
+
+/// Starts profiling the instructions consumed.
+///
+/// Instructions are counted and recorded under the given name until the
+/// `Profile` object returned is dropped.
+pub fn profile(name: &'static str) -> Profile {
+    Profile {
+        name,
+        start_instructions: instruction_count(),
+    }
+}
+
+/// Clears all profiling data.
+pub fn reset() {
+    PROFILING.with(|p| p.borrow_mut().clear());
+}
+
+/// Returns the number of instructions used for each of the profile names.
+pub fn get_results() -> std::collections::BTreeMap<&'static str, u64> {
+    PROFILING.with(|p| p.borrow().clone())
+}
+
+pub struct Profile {
+    name: &'static str,
+    start_instructions: u64,
+}
+
+impl Profile {
+    pub fn new(name: &'static str) -> Self {
+        Self {
+            name,
+            start_instructions: instruction_count(),
+        }
+    }
+}
+
+impl Drop for Profile {
+    fn drop(&mut self) {
+        let instructions_count = instruction_count() - self.start_instructions;
```

Comment:
Borderline paranoid: should this here be a saturating sub instead? It should never happen that the performance counter returns a lower number but if use system time for non wasm architectures you never know what could happen. It'd just be a bit more robust.